src/twisted/web/http.py

```diff
@@ -1985,23 +1987,31 @@ def _dataReceived_CRLF(self) -> bool:
 
     def _dataReceived_TRAILER(self) -> bool:
         """
-        Await the carriage return and line feed characters that follow the
-        terminal zero-length chunk. Then invoke C{finishCallback} and switch to
-        state C{'FINISHED'}.
+        Collect trailer headers if received and finish at the terminal zero-length
+        chunk. Then invoke C{finishCallback} and switch to state C{'FINISHED'}.
 
         @returns: C{False}, as there is either insufficient data to continue,
             or no data remains.
-
-        @raises _MalformedChunkedDataError: when anything other than CRLF is
-            received.
         """
-        if len(self._buffer) < 2:
+        eolIndex = self._buffer.find(b"\r\n", self._start)
+
+        if eolIndex == -1:
+            # Restart the search upon receipt of more data at the start of the
+            # new data, minus one in case the last character of the buffer is
+            # CR.
+            self._start = len(self._buffer) - 1
             return False
 
-        if not self._buffer.startswith(b"\r\n"):
-            raise _MalformedChunkedDataError("Chunk did not end with CRLF")
+        if eolIndex > 0:
+            self._trailerHeaders.append(self._buffer[0:eolIndex])
```

Comment:
will this append happen without limits?

What would happen if a server returns an infinite stream of headers

```
\r\n0\r\n A LOT OF DATA WITHOUT the CRLFCRLF marker.
```

I guess that we will use all the memory

I think that we should add a limit ... maybe 64K of headers ?

I know that we have a similar limit for the HTTP headers.
So maybe use the same limit for trailing headers as for the normal headers.

Reply:
Current headers limit is 16K. I used the same limit for trailer headers.